### DISCUSSION TITLE
'NoneType' object has no attribute 'get_absolute_url'

### DISCUSSION THREAD

**Discussion by f2ed** [netbox-community/netbox#11112] [Q&A] [answered]
I'm currently experiencing this [issue](https://github.com/netbox-community/netbox/issues/11055) and wonder if anyone has any insight in how to figure out the issue.  The issue started after upgrading to 3.3.6, and has persisted through each subsequent upgrade from 3.3.7,8, and 9.  I've tried a new install on a new server by doing a fresh install and importing the database with the same results.  I haven't started with a clean database, as that seems like it would work, and I'm also not sure that I'd be able recreate the steps to reproduce the issue.

**Comment by markkuleinio** [comment, marked as answer]
Now, this is just a wild guess, but if I understood the error location correctly, my guess is that you somehow have IP address object(s) that have inconsistent `assigned_object_id` and/or `assigned_object_type` values. I see the error location (https://github.com/netbox-community/netbox/blob/db7590df1a797de143e9128d880df1a3553c255d/netbox/ipam/tables/ip.py#L378) was added in 3.3.6 (https://github.com/netbox-community/netbox/commit/53c9c3cf8dbe96f6e8b6f40a14c71a52cb2247f6) so the problem presented itself at that point.

I, for one, had that kind of data problems for some IP objects a long time ago, I detected and fixed them using the API.

**Reply by markkuleinio** [reply]
FWIW, https://github.com/netbox-community/netbox/discussions/9259#discussioncomment-2708773 is my earlier comment about that problem.

**Reply by f2ed** [reply]
Thanks.  That's helpful. I will start there.

**Reply by f2ed** [reply]
This is the root of the issue, but it proving to be difficult to track down exactly which IP's seem to be the problem.  I've gone through and found some data integrity issues and corrected those.  But some issues still persist.  And IP's won't display.  I feel like its because the related IP's that are listed on the page are creating the problem, and not the IP I'm actually trying to view.

**Reply by markkuleinio** [reply]
To find those addresses is easy with API. Example in Python:

```
import pynetbox

netbox = pynetbox.api("https://your.netbox.server", token="xxxxxxx", threading=True)
all_ip_addresses = netbox.ipam.ip_addresses.filter()
print("IP addresses:", len(all_ip_addresses))
count = 0
for ip in all_ip_addresses:
    if not ip.assigned_object and (ip.assigned_object_type or ip.assigned_object_id):
        print(
            "assigned_object_type is set to", ip.assigned_object_type,
            "and assigned_object_id is set to", ip.assigned_object_id,
            "(but assigned_object is null) in IP address", ip.address,
        )
    count += 1
    if count % 100 == 0:
        print(count)
```

**Reply by candlerb** [reply]
Also as temporary workaround while you sort out the data issues:

```
--- netbox/ipam/tables/ip.py	2023-04-27 17:03:16.000000000 +0100
+++ netbox/ipam/tables/ip.py.new	2023-05-02 10:18:32.490395733 +0100
@@ -380,7 +380,7 @@
     )
     assigned = columns.BooleanColumn(
         accessor='assigned_object_id',
-        linkify=lambda record: record.assigned_object.get_absolute_url(),
+        linkify=lambda record: record.assigned_object.get_absolute_url() if record.assigned_object else None,
         verbose_name='Assigned'
     )
     comments = columns.MarkdownColumn()
```

**Reply by arnaud-asp** [reply]
I have the same error, but the script proposed markkuleinio is not returning any ip's to me. 

strangely error happens inconsistently, I can't reproduce the error with my user meaning I can browse /ipam/ip-addresses/ fine but two of my colleagues consistently have the error : 

<class 'AttributeError'>
'NoneType' object has no attribute 'get_absolute_url'
Python version: 3.9.13
NetBox version: 3.5.0

It's not cache or browser related as when I log in on their laptop with my user, the issue is not present.
Unfortunately not sure when it started or what caused this as I am usually looking for ip's from /ipam/prefixes and I dont seem to be impacted.

Happy to hear any suggestion on how I could solve this if you have one

**Reply by markkuleinio** [reply]
What is the page length configured for those users? Maybe you have shorter page length and thus the problem does not occur right away?

**Reply by arnaud-asp** [reply]
I have 1000, my colleague has 25, he changed to 1000 (doesnt make sense here but just to try) and he still has the error

**Reply by markkuleinio** [reply]
What other reasons you could think why you don't see the problematic IP address objects but your colleagues see?

**Reply by kkthxbye-code** [reply]
They might have other columns on the table than you, try to reset both yours and theirs to the default columns.

**Reply by candlerb** [reply]
> I have the same error, but the script proposed markkuleinio is not returning any ip's to me.

If you have a large number of IPs, I'm not sure if the Netbox API will hit on paging limits (i.e. that script might not see all IPs).  Try a direct SQL query instead:

```
sudo -u postgres psql netbox

select id,address,assigned_object_type_id,assigned_object_id from ipam_ipaddress
where (assigned_object_type_id is not null and assigned_object_id is null)
or (assigned_object_type_id is null and assigned_object_id is not null);
```

**Reply by arnaud-asp** [reply]
indeed, as soon as I reset my columns I have the issue, this is my settings prior to resetting them :

![image](https://user-images.githubusercontent.com/127384402/235858919-16da1a1c-f476-4ad6-b021-87315c251eab.png)

And below the logs I found after resetting my settings :

May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]: Internal Server Error: /ipam/ip-addresses/
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]: Traceback (most recent call last):
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/core/handlers/exception.py", line 56, in inner
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    response = get_response(request)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/core/handlers/base.py", line 197, in _get_response
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    response = wrapped_callback(request, *callback_args, **callback_kwargs)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/views/generic/base.py", line 103, in view
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return self.dispatch(request, *args, **kwargs)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/netbox/netbox/views/generic/base.py", line 77, in dispatch
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return super().dispatch(request, *args, **kwargs)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/netbox/utilities/views.py", line 99, in dispatch
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return super().dispatch(request, *args, **kwargs)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/views/generic/base.py", line 142, in dispatch
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return handler(request, *args, **kwargs)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/netbox/netbox/views/generic/bulk_views.py", line 177, in get
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return render(request, self.template_name, context)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/shortcuts.py", line 24, in render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    content = loader.render_to_string(template_name, context, request, using=using)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/loader.py", line 62, in render_to_string
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return template.render(context, request)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/backends/django.py", line 61, in render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return self.template.render(context)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/base.py", line 175, in render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return self._render(context)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/test/utils.py", line 111, in instrumented_test_render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return self.nodelist.render(context)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/base.py", line 1005, in render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return SafeString("".join([node.render_annotated(context) for node in self]))
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/base.py", line 1005, in <listcomp>
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return SafeString("".join([node.render_annotated(context) for node in self]))
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/base.py", line 966, in render_annotated
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return self.render(context)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/loader_tags.py", line 157, in render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return compiled_parent._render(context)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/test/utils.py", line 111, in instrumented_test_render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return self.nodelist.render(context)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/base.py", line 1005, in render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return SafeString("".join([node.render_annotated(context) for node in self]))
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/base.py", line 1005, in <listcomp>
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return SafeString("".join([node.render_annotated(context) for node in self]))
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/base.py", line 966, in render_annotated
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return self.render(context)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/loader_tags.py", line 157, in render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return compiled_parent._render(context)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/test/utils.py", line 111, in instrumented_test_render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return self.nodelist.render(context)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/base.py", line 1005, in render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return SafeString("".join([node.render_annotated(context) for node in self]))
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/base.py", line 1005, in <listcomp>
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return SafeString("".join([node.render_annotated(context) for node in self]))
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/base.py", line 966, in render_annotated
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return self.render(context)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/loader_tags.py", line 63, in render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    result = block.nodelist.render(context)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/base.py", line 1005, in render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return SafeString("".join([node.render_annotated(context) for node in self]))
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/base.py", line 1005, in <listcomp>
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return SafeString("".join([node.render_annotated(context) for node in self]))
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/base.py", line 966, in render_annotated
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return self.render(context)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/loader_tags.py", line 63, in render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    result = block.nodelist.render(context)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/base.py", line 1005, in render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return SafeString("".join([node.render_annotated(context) for node in self]))
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/base.py", line 1005, in <listcomp>
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return SafeString("".join([node.render_annotated(context) for node in self]))
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/base.py", line 966, in render_annotated
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return self.render(context)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/loader_tags.py", line 208, in render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return template.render(context)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/base.py", line 177, in render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return self._render(context)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/test/utils.py", line 111, in instrumented_test_render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return self.nodelist.render(context)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/base.py", line 1005, in render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return SafeString("".join([node.render_annotated(context) for node in self]))
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/base.py", line 1005, in <listcomp>
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return SafeString("".join([node.render_annotated(context) for node in self]))
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/base.py", line 966, in render_annotated
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return self.render(context)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/defaulttags.py", line 541, in render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return self.nodelist.render(context)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/base.py", line 1005, in render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return SafeString("".join([node.render_annotated(context) for node in self]))
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/base.py", line 1005, in <listcomp>
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return SafeString("".join([node.render_annotated(context) for node in self]))
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/base.py", line 966, in render_annotated
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return self.render(context)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django_tables2/templatetags/django_tables2.py", line 166, in render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return template.render(context={"table": table}, request=request)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/backends/django.py", line 61, in render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return self.template.render(context)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/base.py", line 175, in render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return self._render(context)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/test/utils.py", line 111, in instrumented_test_render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return self.nodelist.render(context)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/base.py", line 1005, in render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return SafeString("".join([node.render_annotated(context) for node in self]))
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/base.py", line 1005, in <listcomp>
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return SafeString("".join([node.render_annotated(context) for node in self]))
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/base.py", line 966, in render_annotated
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return self.render(context)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/defaulttags.py", line 238, in render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    nodelist.append(node.render_annotated(context))
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/base.py", line 966, in render_annotated
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return self.render(context)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django/template/defaulttags.py", line 193, in render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    values = list(values)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django_tables2/rows.py", line 244, in items
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    column.current_value = self.get_cell(column.name)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django_tables2/rows.py", line 197, in get_cell
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return self._get_and_render_with(
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django_tables2/rows.py", line 173, in _get_and_render_with
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return render_func(bound_column, value)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django_tables2/rows.py", line 208, in _call_render
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return bound_column.link(content, **render_kwargs) if bound_column.link else content
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django_tables2/columns/base.py", line 153, in __call__
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    attrs = self.get_attrs(**kwargs)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django_tables2/columns/base.py", line 148, in get_attrs
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    attrs["href"] = self.compose_url(**kwargs)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django_tables2/columns/base.py", line 100, in compose_url
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return call_with_appropriate(self.url, kwargs)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/venv/lib64/python3.9/site-packages/django_tables2/utils.py", line 572, in call_with_appropriate
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    return fn(**kwargs)
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:  File "/opt/netbox/netbox/ipam/tables/ip.py", line 383, in <lambda>
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]:    linkify=lambda record: record.assigned_object.get_absolute_url(),
May  3 09:46:54 BH-NETBOX-01 gunicorn[1464]: AttributeError: 'NoneType' object has no attribute 'get_absolute_url'

**Reply by arnaud-asp** [reply]
Indeed @candlerb : the sql query got me the 3 culprit ip's, thank you very much everyone for the support!

**Reply by candlerb** [reply]
If that script doesn't find anything, then possibly you have an assigned_object_type_id and assigned_object_type which are both not null, but the combination is pointing to a non-existent object. Those are a bit trickier to find: there is a query at the bottom of https://github.com/netbox-community/netbox/discussions/11726#discussioncomment-4937735 which might help.

Also two more references:

* Custom script to fix addresses where one is null: https://github.com/netbox-community/customizations/pull/85/files
* Try the workaround here: https://github.com/netbox-community/netbox/discussions/11112#discussioncomment-5778670

Using a variation of that workaround which (for example) raises an exception containing the IP address's primary key might identify the problematic IP address(es).

**Reply by markkuleinio** [reply]
I updated the API example code above to catch the other faulty condition as well.

**Reply by candlerb** [reply]
> elif ip.assigned_object and not ip.assigned_object_type:

I don't think what you've written will work:

```
elif ip.assigned_object and not ip.assigned_object_type:
```

ip.assigned_object is an object; it will be None if *either* assigned_object_type_id or assigned_object_id is None. That's essentially what the actual problem is: `record.assigned_object.get_absolute_url()` fails because `record.assigned_object` is None. That means the condition above is never going to be true.

I think the whole thing just needs to be:

```
if ip.assigned_object is None and (ip.assigned_object_type_id or ip.assigned_object_id):
```

I might be wrong though: I'm thinking in terms of the Django ORM models, not in terms of REST API responses.

**Reply by markkuleinio** [reply]
Fair enough, edited the code again.

**Comment by PieterL75** [comment]
Just experienced the same thing here.. 
3 IP addresses that did not have an 'assigned_object', even though there was an 'assing_object_id' and 'assigned_object_type'
Looks like the IP was not assigned to an interface, but it was still 'primary ip' for that device

**Comment by bluikko** [comment]
Have this error also right now, after upgrading from 3.5.8 to 3.7.8 and then to 4.0.6.  
This is an instance that has been upgraded from early v2 onwards.

SQL query in https://github.com/netbox-community/netbox/discussions/11112#discussioncomment-5788351 gets 0 rows.
Running the script in https://github.com/netbox-community/netbox/discussions/11112#discussioncomment-4375550 shows that **all** the thousands of IP addresses are invalid:
```
assigned_object_type is set to dcim.interface and assigned_object_id is set to 8733 (but assigned_object is null) in IP address 192.0.2.1/26
```

So I assume this could be fixed with a script since we do have at least `assigned_object_id`?

Somehow the upgrade database migration has messed up each and every IP address in the database.  
I am very surprised that only a small amount of users have been hit with this problem.

**Reply by candlerb** [reply]
The condition says:

```
if not ip.assigned_object and (ip.assigned_object_type or ip.assigned_object_id):
```

Given that assigned_object is None, but the log message says that *both* of the other fields are set (i.e. dcim.interface and 8733), I think it means that you have no interface with ID 8733. Check at `/dcim/interfaces/8733/`

If that's true, then maybe the interface has vanished at some point in the past; or it means that either the assigned_object_type or the assigned_object_id has the wrong value. For example, do you have a virtualization.vminterface with ID 8733? Check at `/virtualization/interfaces/8733/`.

If both of those are missing, then can you you find any device or VM which has this IP address as its primary IP?

If not, the simplest way to clear this error would be to set both assigned_object_type and assigned_object_type to null on the IP address. Afterwards you can decide if you want to assign it to an interface, and if so which one.

**Reply by bluikko** [reply]
Heh you answered so quickly I did not even have time to look much into this. Thanks for that.

By the way I forgot to say that all devices/VMs show correct IPv4/6 primary IP addresses. I noticed this problem first after update because primary IP addresses were listed but could not be edited at all (and showed `----` when editing) and interface list showed no IP addresses.

The `assigned_object_type` is correct `dcim.interface` for at least the 1 example IP address and that device interface does exist and is even the right one.

In other words: seems `assigned_object_type` and `assigned_object_id` are correct and just the `assigned_object` is `none`. It might be easy to fix then.

Edited to add: I want to stress that **all** the thousands of IP addresses are in this state.

**Reply by candlerb** [reply]
> In other words: seems assigned_object_type and assigned_object_id are correct and just the assigned_object is none. It might be easy to fix then.

If the data is valid, then that sounds like something broken in Django (or broken by a plugin).  What happens if you try one of the affected interfaces in nbshell? It works fine for me with 4.0.6:

```
# /opt/netbox/venv/bin/python /opt/netbox/netbox/manage.py nbshell
### NetBox interactive shell (netbox4)
### Python 3.12.3 | Django 5.0.6 | NetBox 4.0.6
### lsmodels() will show available models. Use help(<model>) for more info.
>>> ip = IPAddress.objects.get(address="10.12.0.1/24")
>>> ip.pk
273
>>> ip.assigned_object
<Interface: vlan256>
>>> ip.assigned_object.device
<Device: gw2>
```

And then cross-check the data in SQL:

```
netbox=# select id,assigned_object_type_id,assigned_object_id from ipam_ipaddress where id=273;
 id  | assigned_object_type_id | assigned_object_id
-----+-------------------------+--------------------
 273 |                      17 |                448
(1 row)

netbox=# select * from django_content_type where id=17;
 id | app_label |   model
----+-----------+-----------
 17 | dcim      | interface
(1 row)

netbox=# select id,name from dcim_interface where id=448;
 id  |  name
-----+---------
 448 | vlan256
(1 row)
```

**Reply by bluikko** [reply]
I do not have any plugins. And never had any plugins either, only stock NetBox.

But, modifying your script I do have some new data.

1% of IP addresses have `none` set on all `assigned_object_type`, `assigned_object_id` and `assigned_object`.  
I will call the above 1% _category B_ and the initial 99% problem IP addresses as _category A_.

All category B IP addresses are in just 3 different prefixes so they are easy to fix manually.  
These 3 prefixes have a custom validator (which has existed since around the custom validator feature was released).  
Also these same 3 prefixes also have a very, very small amount of category A IP addresses:  
- In these prefixes the category A IP addresses are *very old*, probably before the custom validator existed.
- And category B IP addresses seem to be added after that custom validator was configured.

I believe this might be related to the bug with custom validators and some attributes, the details escape me now but I might look for it.

It looks clear from the split, that having a custom validator -- or being upgraded from/to a version around that time -- has somehow broken all IP addresses in 2 different ways:
- One type of broken IP addresses before that point in time, in prefixes that had custom validator.
- And all later added IP addresses with custom validator, and IP addresses that didn't have custom validators, broken in a different way.

Edited: important fix at last sentence.

**Reply by candlerb** [reply]
I don't really get it. You're saying 99% of IP addresses (category A) have values set for assigned_object_type and assigned_object_id, but not assigned_object.

However assigned_object is just an accessor that gives you the object that is referred to by assigned_object_type_id and assigned_object_id. Therefore, this should be impossible.

You need to provide an example of one of these addresses, with both the SQL output and the nbshell output as I showed before.

**Reply by bluikko** [reply]
Aha! I didn't know it is just an accessor.

That lead me to disable the custom validator and everything works.  
I guess something has changed in custom validators between 3.5 and 4.0, don't recall seeing anything in READMEs (but then again, the API version authentication requirement is not in READMEs either........)

Thank you for your time, and I apologize for kinda wasting it!

**Reply by candlerb** [reply]
Not at all. And if you can provide an example of how *not* to write a validator - i.e. a validator which messes things up in this way - that would be very useful to help others avoid the same problem. It might even be a bug in Netbox's interface for validators.

**Reply by bluikko** [reply]
I will definitely update after I get to the bottom of this.

Everything worked fine in 3.5.8. And going back to when validators released.

**Reply by bluikko** [reply]
Unfortunately it is not something simple with my custom validator.

I first turned back on all custom validators from `configuration.py` and commented out everything in the "suspected problem" validator function.  
I enabled the validator function in small pieces and everything continued to work. The validator *and* configuration file are now exactly like they were before and everything continues to work. Validator works as expected and all IP addresses work.

I wonder if the validator could somehow interfere with upgrades?  
After upgrade I did test that the validator worked: I edited one IP address through `Prefixes` to invalid data and checked that validator rejected the change, canceled out from the changes.  
Only after this I checked any IP addresses, either assigned in `Device` `Interfaces` or `IP Addresses` table and nothing related to them worked.

Looking at all of this, I noticed that I did not restart the netbox/rq services after restoring the database from another 4.0.6 instance.  
_First restart was after disabling the validator_. Which was dumb.
Perhaps this is the real reason. don't think I will have anything else to add to this issue, I have gone through everything.

**Reply by candlerb** [reply]
If it's just a question of forgetting to restart after modifying the database, then that's pretty clear. It would have been nice to see the nbshell and sql queries while the problem was still manifesting itself though. They probably would have shown everything was fine (since nbshell is its own process) and thus pointed the finger at something wrong with the webservice, like not being restarted.

**Comment by lono0778** [comment]
Hello,

actually i have a similar problem after upgrade to netbox 4.0.6 from 3.7.0.

When i modify a permission whit a JSON value i get the following error after clicking  save:

Server Error

There was a problem with your request. Please contact an administrator.

The complete exception is provided below:

<class 'AttributeError'>

'NoneType' object has no attribute 'objects'

Python version: 3.10.12
NetBox version: 4.0.6
Plugins: 
  netbox_dns: 1.0.4
  netbox_documents: 0.7.0
  netbox_floorplan: 0.4.0
  netbox_qrcode: 0.0.13

Any suggestions here?

Thank you
